File: leetcode3/황은지/1004. Max Consecutive Ones III.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function (nums, k) {
  let start = 0;
  let count = 0;
  let maxLen = 0;

  for (let end = 0; end < nums.length; end++) {
    if (nums[end] === 0) count++;

    while (count > k) {
      if (nums[start++] === 0) count--;
    }
    maxLen = Math.max(maxLen, end - start + 1);
  }
  return maxLen;
};

File: leetcode3/남효정/1004. Max Consecutive Ones III.py

'''
1. 아이디어 :
2. 시간복잡도 :
3. 자료구조/알고리즘 :
'''

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        left = 0
        zero_cnt = 0
        max_len = 0

        for right in range(len(nums)):
            if nums[right] == 0:
                zero_cnt += 1

            while zero_cnt > k:
                if nums[left] == 0:
                    zero_cnt -= 1
                left += 1

            max_len = max(max_len, right - left + 1)

        return max_len
        